File: 백준/Silver/1389. 케빈 베이컨의 6단계 법칙/케빈 베이컨의 6단계 법칙.java
import java.io.*;
import java.util.*;

public class Main {
    static class Node {
        int x, cnt;
        public Node(int x, int cnt) {
            this.x = x;
            this.cnt = cnt;
        }
    }
    public static void main(String[] args) throws IOException {
        BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
        StringTokenizer st = new StringTokenizer(br.readLine());
        int N = Integer.parseInt(st.nextToken());
        int M = Integer.parseInt(st.nextToken());
        ArrayList<Integer>[] arr = new ArrayList[N + 1];
        int min = Integer.MAX_VALUE;
        int min_idx = 0;
        for(int i = 1; i <= N; i++) {
            arr[i] = new ArrayList<Integer>();
        }
        for(int i = 0; i < M; i++) {
            st = new StringTokenizer(br.readLine());
            int a = Integer.parseInt(st.nextToken());
            int b = Integer.parseInt(st.nextToken());
            arr[a].add(b);
            arr[b].add(a);
        }
        for(int i = 1; i <= N; i++) {
            Queue<Node> que = new LinkedList<>();
            boolean[] visited = new boolean[N + 1];
            que.add(new Node(i, 0));
            visited[i] = true;
            int sum = 0;
            while(!que.isEmpty()) {
                Node node = que.poll();
                sum += node.cnt;
                for(int num : arr[node.x]) {
                    if(!visited[num]) {
                        que.add(new Node(num, node.cnt + 1));
                        visited[num] = true;
                    }
                }
            }
            if(sum < min) {
                min = sum;
                min_idx = i;
            }
        }
        System.out.println(min_idx);
    }
}

File: 백준/Silver/1389. 케빈 베이컨의 6단계 법칙/케빈 베이컨의 6단계 법칙.js
class Queue {
  constructor() {
    this.arr = [];
    this.head = 0;
  }

  add(num) {
    this.arr.push(num);
  }

  poll() {
    return this.arr[this.head++];
  }

  isEmpty() {
    return this.arr.length === this.head;
  }
}

const [[N, M], ...input] = require('fs').readFileSync('/dev/stdin').toString().trim().split('\n').map(line => line.trim().split(' ').map(Number));
const arr = Array.from({ length: N + 1 }, () => []);
let min = Number.MAX_VALUE;
let min_idx = 0;
for(let i = 0; i < M; i++) {
  arr[input[i][0]].push(input[i][1]);
  arr[input[i][1]].push(input[i][0]);
}
for(let i = 1; i <= N; i++) {
  const que = new Queue();
  const visited = new Array(N + 1).fill(false);
  que.add({ x : i, cnt : 0});
  visited[i] = true;
  let sum = 0;
  while(!que.isEmpty()) {
    const node = que.poll();
    sum += node.cnt;
    for(let num of arr[node.x]) {
      if(!visited[num]) {
        que.add({x : num, cnt : node.cnt + 1});
        visited[num] = true;
      }
    }
  }
  if(sum < min) {
    min = sum;
    min_idx = i;
  }
}
console.log(min_idx);